### ISSUE TITLE
Incorrect order: conv-BN-ReLU instead of BN-ReLU-conv

### ISSUE THREAD

**Issue by apacha** [titu1994/Wide-Residual-Networks#4]
Described in the original paper and as can be seen in the [original implementation](https://github.com/szagoruyko/wide-residual-networks/blob/master/models/wide-resnet.lua#L57), the authors suggested that for blocks, the internal structure should be: `BatchNormalization-ReLU-Convolution` instead of `Convolution-BatchNormalization-ReLU`. 

Any particular reason, why the order was changed? If not, I would propose that we change it to represent the same architecture as proposed by the original authors.

**Comment by titu1994** [comment]
It is following batch norm - relu - conv.

Please look at the diagram provided in the readme. Initial conv at the start, followed by batch norm and then activation. This is then passes to the two branches. 

In essence, the branch initial conv begins with batchnorm-relu-conv and then another block. 

The implementation is correct.

**Comment by apacha** [comment]
I didn't mean any offense. I am very thankful for this implementation, especially, since it is already updated to Keras 2. I just compared your architecture

![2017-05-29 17_55_35-titu1994_wide-residual-networks_ wide residual networks in keras](https://cloud.githubusercontent.com/assets/594497/26555969/0d13ed86-4498-11e7-8361-8a45cca58b43.png)

to the following implementation https://github.com/asmith26/wide_resnets_keras 

![2017-05-29 17_55_03-asmith26_wide_resnets_keras_ keras implementation pretrained weights for _wide](https://cloud.githubusercontent.com/assets/594497/26555970/0d150afe-4498-11e7-98ef-2586efa336bd.png)

that definitely does look different between two merge-blocks. 

I am just trying to figure out, what version performs better. Clearly, there is no right or wrong, but since the authors explicitly mentioned, that they changed from conv-BN-ReLU to BN-ReLU-conv, I was confused at which part, they changed this order. Unfortunately, the [Lua code](https://github.com/szagoruyko/wide-residual-networks/blob/master/models/wide-resnet.lua) does not reveal that information to me easily.

**Comment by titu1994** [comment]
Hmm, I agree that it looks different. 

This is because of the 1st merge layer. The other repo performs the second conv and then immediately merges with the parallel conv. 

I perform batch norm and activation and then merge. That is the reason for the difference in architecture. 

Hmm, if I had to look at it critically, the architecture works in both forms, but the other repository is closer to the original paper. 

I can definitely shuffle the order, but that would mean the trained models will have to be dropped and new models need to be trained. 

Currently I am focused on another model for research, so I can't allocate resources to train a WRN at this time. However, being close to the original paper is also important. I will have to think about changing the model building code.

**Comment by apacha** [comment]
Glad, that you see the difference. Currently I am trying both implementations on CIFAR-10 with WRN-28-10 to see which one performs better. I expect results by tomorrow later during the day.

If it is just about computational power, I can run the updated script on my machine in the next available slot and upload the weights, once the code has been updated and you are happy with the implementation.

**Comment by titu1994** [comment]
Upon further looking at it, it seems that the difference in the architecture doesn't account for a performance loss. Considering the other repo has provided weights for WRN-40-4 and has obtained 7 % error, it is in line with when I first tried 40-4 and got 6.97% error.

I don't think I will be changing the model building script right now, since the performance difference seems to be negligible at first glance. 

However, if your test of both models shows that their model infact performs better, I will quickly change the model building scripts to be in line with the paper.

**Comment by apacha** [comment]
I've been running the trainings for both WRN-28-10 networks twice and the results indicate that the other network clearly outperformed this implementation approximately by 1%. This would suggest that the originally proposed architecture is more suited for this task and you might want to change this implementation.

Results
----
This architecture:
First run (weight-decay 0.0005): 92.9%
Second run (weight-decay 0.0001): 93.4%

Other architecture:
First run (weight-decay 0.0005): 94.2%
Second run (weight-decay 0.0001): 94.5%

Training details
----
Note a couple of things:

- I've used Adadelta for both trainings, as SGD often failed totally, leading to vanishing gradient or exploding gradient.
- I've only used 40000 images for the training, 10000 for validation and 10000 for testing. So I assume that the results could be much better, if 50k/10k split is used without validation set
- I've used data-augmentation with 20% random zoom and 10° random rotation
- I've used dynamic learning rate reduction upon reaching a plateau (8 epochs no improvement) by factor 0.7
- I've used early stopping to terminate after 20 epochs of no improvement
